### PR TITLE
Code refactoring smart_resolve_report_errors

### PR DESCRIPTION
**Pull request by chenyukang**
`smart_resolve_report_errors` https://github.com/rust-lang/rust/blob/4ecfdfac51b159f68fce608792affb34a70e6f73/compiler/rustc_resolve/src/late/diagnostics.rs#L143
is almost 600 lines of code, we should do some code refactoring.

### CODE REVIEW

**Comment by chenyukang** [comment]
Original thoughts from 
https://github.com/rust-lang/rust/pull/101908#issuecomment-1251187688

File: compiler/rustc_resolve/src/late/diagnostics.rs [hunk 1]
```diff
@@ -277,53 +270,92 @@ impl<'a: 'ast, 'ast> LateResolutionVisitor<'a, '_, 'ast> {
...
+                    }
                 }
             }
-            _ => {}
         }
+    }
 
+    fn suggest_self_or_self_ref(&mut self, err: &mut Diagnostic, path: &[Segment], span: Span) {
         let is_assoc_fn = self.self_type_is_available();
+        let item_str = path.last().unwrap().ident;
```
Comment (cjgillot) [inline compiler/rustc_resolve/src/late/diagnostics.rs:358]:
```suggestion
        let Some(path_last_segment) = path.last() else { return };
        let item_str = path_last_segment.ident;
```
Let's avoid `unwrap`s in diagnostic code.  It's a cheap way to avoid accidental ICEs.

File: compiler/rustc_resolve/src/late/diagnostics.rs [hunk 2]
```diff
@@ -724,38 +710,106 @@ impl<'a: 'ast, 'ast> LateResolutionVisitor<'a, '_, 'ast> {
                 }
             }
         }
-
-        (err, candidates)
     }
 
-    fn detect_assoct_type_constraint_meant_as_path(&self, base_span: Span, err: &mut Diagnostic) {
```
Comment (cjgillot) [inline compiler/rustc_resolve/src/late/diagnostics.rs:731]: Why did you move this method?  This makes review more complicated.

File: compiler/rustc_resolve/src/late/diagnostics.rs [hunk 3]
```diff
@@ -724,38 +710,106 @@ impl<'a: 'ast, 'ast> LateResolutionVisitor<'a, '_, 'ast> {
...
-                };
-                for bound in bounds {
-                    let ast::GenericBound::Trait(trait_ref, ast::TraitBoundModifier::None)
-                        = bound else
-                    {
-                        continue;
-                    };
-                    if base_span == trait_ref.span {
+    /// Emit special messages for unresolved `Self` and `self`.
+    fn suggest_self_ty_or_self_value(
```
Comment (cjgillot) [inline compiler/rustc_resolve/src/late/diagnostics.rs:716]: Could you separate the `self_type` and `self_value` cases?

File: compiler/rustc_resolve/src/late/diagnostics.rs [hunk 4]
```diff
@@ -358,96 +390,25 @@ impl<'a: 'ast, 'ast> LateResolutionVisitor<'a, '_, 'ast> {
...
-                        "`self` not allowed in {} {}",
-                        item_kind.kind.article(),
-                        item_kind.kind.descr()
-                    ),
-                );
-            }
-            return (err, Vec::new());
-        }
-
+    fn try_lookup_name_with_relex_fashion(
```
Comment (cjgillot) [inline compiler/rustc_resolve/src/late/diagnostics.rs:395]:
```suggestion
    fn try_lookup_name_relaxed(
```

File: compiler/rustc_resolve/src/late/diagnostics.rs [hunk 5]
```diff
@@ -579,29 +541,40 @@ impl<'a: 'ast, 'ast> LateResolutionVisitor<'a, '_, 'ast> {
...
+        source: PathSource<'_>,
+        path: &[Segment],
+        res: Option<Res>,
+        span: Span,
+        base_error: &BaseError<'static>,
+    ) {
         let is_macro =
             base_error.span.from_expansion() && base_error.span.desugaring_kind().is_none();
-        if !self.type_ascription_suggestion(&mut err, base_error.span) {
+        if !self.type_ascription_suggestion(err, base_error.span) {
```
Comment (cjgillot) [inline compiler/rustc_resolve/src/late/diagnostics.rs:577]: Could you pull this test outside of the call?

File: compiler/rustc_resolve/src/late/diagnostics.rs [hunk 6]
```diff
@@ -579,29 +541,40 @@ impl<'a: 'ast, 'ast> LateResolutionVisitor<'a, '_, 'ast> {
...
-                self.r.add_typo_suggestion(&mut err, typo_sugg, ident_span);
-                return (err, candidates);
+                self.r.add_typo_suggestion(err, typo_sugg, ident_span);
+                return (true, candidates);
             }
         }
+        return (false, candidates);
+    }
 
+    fn suggest_type_ascription(
```
Comment (cjgillot) [inline compiler/rustc_resolve/src/late/diagnostics.rs:566]: This name is not really describing what this function does.

File: compiler/rustc_resolve/src/late/diagnostics.rs [hunk 7]
```diff
@@ -130,6 +130,16 @@ pub(super) enum LifetimeElisionCandidate {
     Missing(MissingLifetime),
 }
 
+/// Only used for diagnostics.
+struct BaseError<'a> {
```
[the remark this answers is not in the document]
Comment (chenyukang) [reply compiler/rustc_resolve/src/late/diagnostics.rs:134]: You mean `&str` also replaced with `String`?

[the remark this answers is not in the document]
Comment (cjgillot) [reply compiler/rustc_resolve/src/late/diagnostics.rs:134]: No, I mean `&'static str`.

File: compiler/rustc_resolve/src/late/diagnostics.rs (hunk 2 above)
Comment (chenyukang) [reply compiler/rustc_resolve/src/late/diagnostics.rs:731]: This method keep the same except postions difference after code refactored,
https://github.com/chenyukang/rust/blob/code-refactor/compiler/rustc_resolve/src/late/diagnostics.rs#L305

Yes, if most part of the function changed, the diff is not easy to review.

Comment (cjgillot) [reply compiler/rustc_resolve/src/late/diagnostics.rs:731]: Could you make 1 commit per method? Just say so if it's not practical.

Comment (chenyukang) [reply compiler/rustc_resolve/src/late/diagnostics.rs:731]: It's hard now, 
since I didn't make it that way from the starting of refctoring😂

Comment (cjgillot) [reply compiler/rustc_resolve/src/late/diagnostics.rs:731]: Ok, so let's leave it as is :smile:.

File: compiler/rustc_resolve/src/late/diagnostics.rs (hunk 6 above)
Comment (chenyukang) [reply compiler/rustc_resolve/src/late/diagnostics.rs]: Updated!

File: compiler/rustc_resolve/src/late/diagnostics.rs (hunk 5 above)
Comment (chenyukang) [reply compiler/rustc_resolve/src/late/diagnostics.rs]: Yes, that looks better.

File: compiler/rustc_resolve/src/late/diagnostics.rs [hunk 8]
```diff
@@ -277,53 +270,92 @@ impl<'a: 'ast, 'ast> LateResolutionVisitor<'a, '_, 'ast> {
...
+        debug!(?res, ?source);
+        let base_error = self.make_base_error(path, span, source, res);
         let code = source.error_code(res.is_some());
         let mut err =
             self.r.session.struct_span_err_with_code(base_error.span, &base_error.msg, code);
 
         self.suggest_swapping_misplaced_self_ty_and_trait(&mut err, source, res, base_error.span);
 
-        if let Some((span, label)) = base_error.span_label {
+        if let Some((span, label)) = base_error.span_label.clone() {
```
[the remark this answers is not in the document]
Comment (chenyukang) [reply compiler/rustc_resolve/src/late/diagnostics.rs]: This just comes in my mind, is it always better to avoid `.clone()` when the type implemented `Copy`?
If so, maybe we should add a lint rule or diagnositcs for it.....
